### PR TITLE
WMCO Upgrade enhancement design update

### PR DESCRIPTION
**Pull request by VaishnaviHire**
This change reflects the updated implementation of the upgrade process of WMCO.
Previous upgrade design states that WMCO makes use of `maxUnavailable` field in
the Windows MachineSet to determine deletion of nodes with version annotation
mismatch. As this field is not available in the MachineSet spec, the updated
design makes use of `maxUnhealthy` count defined internally by WMCO. This count
ensures that we have only maxUnhealthy number of nodes in Not Ready state
during upgrades

### CODE REVIEW

File: enhancements/windows-containers/windows-machine-config-operator-upgrades.md [hunk 1]
```diff
@@ -60,7 +60,7 @@ will then be able to configure VMs that will be created in place of the
 terminated ones.
 
 In order to allow for this to occur with minimal service disruption, it is
-recommended that users have at least two Windows nodes within the cluster.
+recommended that users have at least three Windows nodes within the cluster.
```
Comment (sdodson) [inline enhancements/windows-containers/windows-machine-config-operator-upgrades.md:63]: What's the reason for this change?

File: enhancements/windows-containers/windows-machine-config-operator-upgrades.md [hunk 2]
```diff
@@ -113,9 +113,12 @@ associated with the node. The [Machine API Operator](https://github.com/openshif
 will drain and delete the node before the Machine deletion completes. It will
 then create a new Machine to reconcile the MachineSet replica count. WMCO will
 configure this new machine and apply the new version annotation. The max amount
-of unavailable nodes at a time will be dependent on the maxUnavailable field
-in the [MachineSet](https://github.com/openshift/machine-api-operator/blob/master/config/machineset.yaml)
-created for Windows machines. WMCO will not render more than the specified
+of unavailable nodes at a time will be dependent on the maxUnhealthy field defined
+internally by the WMCO. This field ensures that we only have maxUnhealthy number of
+nodes in Not Ready state during upgrades. Having limited number of unavailable nodes
```
Comment (sdodson) [inline enhancements/windows-containers/windows-machine-config-operator-upgrades.md:118]: Is this specifically the kubelet `NotReady` status with no space? If so can you correct that?

File: enhancements/windows-containers/windows-machine-config-operator-upgrades.md (hunk 1 above)
Comment (VaishnaviHire) [reply enhancements/windows-containers/windows-machine-config-operator-upgrades.md:63]: With the new upgrade process, we are deleting and reconfiguring nodes one at a time.
Having 3 Windows nodes ensures high availability and avoids any downtime for the workloads.

For eg. when we have only 2 nodes during upgrade, there will be a downtime for Windows workloads, when the workload is scaling up from 0 to 1 on newly upgraded node.

Comment (aravindhp) [reply enhancements/windows-containers/windows-machine-config-operator-upgrades.md:63]: @sdodson should we explicitly call that out?

File: enhancements/windows-containers/windows-machine-config-operator-upgrades.md (hunk 2 above)
Comment (VaishnaviHire) [reply enhancements/windows-containers/windows-machine-config-operator-upgrades.md]: A Windows node is in Not Ready state if it is not configured completely i.e it is missing the version annotation set by WMCO. Updated the description to reflect this.

File: enhancements/windows-containers/windows-machine-config-operator-upgrades.md (hunk 1 above)
Comment (sdodson) [reply enhancements/windows-containers/windows-machine-config-operator-upgrades.md:63]: No, this is probably fine. I was thinking about saying we should say something like "N+1 capacity" but that gets complicated quickly since it's not quite that simple.

I was mostly wondering if this was driven by new critical omponent being added to the Windows stack for OpenShift which must be HA.

Comment (aravindhp) [reply enhancements/windows-containers/windows-machine-config-operator-upgrades.md:63]: >I was mostly wondering if this was driven by new critical omponent being added to the Windows stack for OpenShift which must be HA.

Nothing like that as of now. We just observed downtime on workloads with just two nodes during upgrades and wanted to address that.

File: enhancements/windows-containers/windows-machine-config-operator-upgrades.md (hunk 2 above)
Comment (aravindhp) [reply enhancements/windows-containers/windows-machine-config-operator-upgrades.md]: We shouldn't be using `Not Ready` but rather `not ready` then.

Comment (VaishnaviHire) [reply enhancements/windows-containers/windows-machine-config-operator-upgrades.md]: I have updated this to use `not ready`

**Comment by sdodson** [comment]
/lgtm
/hold
LGTM, didn't know if you were seeking additional review from anyone else, if not feel free to clear the hold.
